libdnf/base/goal.cpp

```diff
@@ -45,8 +45,17 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 namespace {
 
 void add_obsoletes_to_data(const libdnf::rpm::PackageQuery & base_query, libdnf::rpm::PackageSet & data) {
+    // Add only for obsoleters of the latest versions.
+    // This should prevent unexpected results in case a package has multiple versions
+    // and some older version is being obsoleted.
+    // See also https://bugzilla.redhat.com/show_bug.cgi?id=2176263
+    libdnf::rpm::PackageQuery data_query(base_query.get_base(), libdnf::sack::ExcludeFlags::APPLY_EXCLUDES, true);
+    data_query.update(data);
+    data_query.filter_priority();
+    data_query.filter_latest_evr();
```

Comment:
I am thinking about two edge cases where we might have a problem. Both cases are related to situation when package name is installed.
1. When vendor change is not allowed (configuration allow_vendor_change)
2. We will have a different set of candidates in comparison to upgrade request with the same argument

I have a suggestion - if `data` contains installed package than use the original approach. What do you think?

Reply:
I think we should have test scenarios for these situations. I'll try to prepare them - it will definitely help me wrap my head around it.